hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/checksum/ContainerChecksumTreeManager.java

```diff
@@ -143,12 +147,124 @@ public void markBlocksAsDeleted(KeyValueContainerData data, Collection<Long> del
     }
   }
 
-  public ContainerDiff diff(KeyValueContainerData thisContainer, ContainerProtos.ContainerChecksumInfo otherInfo)
+  public ContainerDiff diff(KeyValueContainerData thisContainer, ContainerProtos.ContainerChecksumInfo peerChecksumInfo)
       throws IOException {
-    // TODO HDDS-10928 compare the checksum info of the two containers and return a summary.
-    //  Callers can act on this summary to repair their container replica using the peer's replica.
-    //  This method will use the read lock, which is unused in the current implementation.
-    return new ContainerDiff();
+    Preconditions.assertNotNull(thisContainer, "Container data is null");
+    Preconditions.assertNotNull(peerChecksumInfo, "Peer checksum info is null");
+    Optional<ContainerProtos.ContainerChecksumInfo.Builder> thisContainerChecksumInfoBuilder =
+        read(thisContainer);
```

Comment:
We should add a new `read` that returns `ContainerProtos.ContainerChecksumInfo`, current `read` is implemented to read and then update, so it makes sense to return a builder. Optionally, we can have all the callers of the current `read` create a builder. We can rename `read` to `readBuilder`.. 
Currently, the code flow is a bit odd where we get `ContainerProtos.ContainerChecksumInfo` and then convert it to `Builder` only to convert it back to `ContainerProtos.ContainerChecksumInfo`. This is an unnecessary GC load.